curvefs/src/metaserver/dumpfile.cpp

```diff
@@ -0,0 +1,501 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/**
+ * Project: Curve
+ * Created Date: 2021-07-23
+ * Author: Jingli Chen (Wine93)
+ */
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <sys/prctl.h>
+#include <glog/logging.h>
+
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "src/common/crc32.h"
+#include "src/common/process.h"
+#include "src/common/timeutility.h"
+#include "curvefs/src/metaserver/iterator.h"
+#include "curvefs/src/metaserver/dumpfile.h"
+
+namespace curvefs {
+namespace metaserver {
+
+#define RETURN_IF_UNSUCCESS(statement) \
+do { \
+    auto rc = statement; \
+    if (rc != DUMPFILE_ERROR::OK) { \
+        return rc; \
+    } \
+} while (0)
+
+#define EXIT_LOAD_IF_UNEXPECT(unexpect, status) \
+do { \
+    if ((unexpect) == true) { \
+        isValid_ = false; \
+        dumpfile_->SetLoadStatus(status); \
+        return; \
+    } \
+} while (0)
+
+using ::curve::common::CRC32;
+
+const std::string DumpFile::kCurvefs_ = "CURVEFS";  // NOLINT
+const uint8_t DumpFile::kVersion_ = 1;
+const uint32_t DumpFile::kMaxStringLength_ = 1024 * 1024;  // 1MB
+
+DumpFile::DumpFile(const std::string& pathname)
+    : pathname_(pathname),
+      fd_(-1),
+      fs_(Ext4FileSystemImpl::getInstance()),
+      loadStatus_(DUMPFILE_LOAD_STATUS::INCOMPLETE) {}
+
+DUMPFILE_ERROR DumpFile::Open() {
+    if (fd_ >= 0) {
+        return DUMPFILE_ERROR::OK;
+    }
+
+    auto retCode = fs_->Open(pathname_, O_RDWR | O_CREAT);
```

Comment:
Do you missing O_TRUNC in open flags?
But，actually you should create a temporary file, and dump data into this file, and after those tasks has been done, rename temporary file to pathname_.
Because when you truncate the file, and the process exit abnormaly before dump all data into this file, data will be lost.

Reply:
> Do you missing O_TRUNC in open flags?
> But，actually you should create a temporary file, and dump data into this file, and after those tasks has been done, rename temporary file to pathname_.
> Because when you truncate the file, and the process exit abnormaly before dump all data into this file, data will be lost.

TBD